exchange/bitswap/workers.go

```diff
@@ -215,35 +274,20 @@ func (bs *Bitswap) providerQueryManager(ctx context.Context) {
 			default:
 			}
 
-			activeLk.Lock()
+			ksetLk.Lock()
 			if kset.Has(e.Cid) {
-				activeLk.Unlock()
+				ksetLk.Unlock()
 				continue
 			}
 			kset.Add(e.Cid)
-			activeLk.Unlock()
-
-			go func(e *blockRequest) {
-				child, cancel := context.WithTimeout(e.Ctx, providerRequestTimeout)
-				defer cancel()
-				providers := bs.network.FindProvidersAsync(child, e.Cid, maxProvidersPerRequest)
-				wg := &sync.WaitGroup{}
-				for p := range providers {
-					wg.Add(1)
-					go func(p peer.ID) {
-						defer wg.Done()
-						err := bs.network.ConnectTo(child, p)
-						if err != nil {
-							log.Debug("failed to connect to provider %s: %s", p, err)
-						}
-					}(p)
-				}
-				wg.Wait()
-				activeLk.Lock()
-				kset.Remove(e.Cid)
-				activeLk.Unlock()
-			}(e)
+			ksetLk.Unlock()
 
+			select {
```

Comment:
this will still block callers if the semaphore fills up. Take a look at the providesCollector code, I have an expandable buffer to help avoid those issues